### DISCUSSION TITLE
Leverage more kt compiler plugin API

### DISCUSSION THREAD

**Discussion by picklebento** [detekt/detekt#5958] [General]
We are currently using `AnalysisHandlerExtension` kt compiler plugin API which may involve heavy and unnecessary computation. There are actually more Kotlin compiler plugin API that allows us to handle styling errors more idiomatically (and potentially more effectively)
- [DeclarationChecker](https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/resolve/checkers/DeclarationChecker.kt)
- [CallChecker](https://github.com/JetBrains/kotlin/blob/29b23e79f32791e456a5b4a453277f0f0b3e984d/compiler/frontend/src/org/jetbrains/kotlin/resolve/calls/checkers/CallChecker.kt)

To implement autoCorrect, we can also leverage [IrGenerationExtension](https://github.com/JetBrains/kotlin/blob/master/compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/extensions/IrGenerationExtension.kt), with my bet that this JVM/IR will be stable in 1.5.

This issue is meant for discussion and prototyping, not really an issue to be addressed soon.

**Comment by schalkms** [comment]
FYI: I converted this to a discussion.
